feat(component): Add Text-input component (#3093)
Orange-OpenSource/Orange-Boosted-Bootstrap: site/src/content/docs/forms/text-input.mdx

---
title: Text input

> description: A text input is a user interface component that allows users to enter, edit, or select single-line textual data. It provides a visual and interactive affordance for text entry while supporting labels, placeholders, icons, helper messages, and validation feedback.

Reviewer: Does it really allow to "select single-line textual data"? This make me think of a `select input`... I'm not sure this is necessary

aliases:
  - "/docs/forms/text-input/"
toc: true
---

import { getConfig } from '@libs/config'

<Callout type="info">
{/* TODO: set the right url for design documentation */}
You can find here the [OUDS Text Input design guidelines](https://unified-design-system.orange.com/).
</Callout>

## Overview

> Wrap a pair of `<input class="text-input-field">` and `<label>` elements in `.text-input-container` to style OUDS textual form fields. Note that the `<label>` must come before the `<input>`.

Reviewer: We are not consistent on th `input` style vs `<input>` elsewhere in the documentation but this might be better? I think we should be consistent and switch to this version everywhere if everyone is ok.

Author: Ok for me

Reviewer 2: angled bracket are already used in other documentation to talk about the HTML element, we can stick to that notation.

Author: Ok so we can solved this one?

A `placeholder` attribute is required on each `<input>` as our CSS-only floating label implementation relies on the `:placeholder-shown` pseudo-element to detect when the input is empty.

If the placeholder text is empty, it is not visible; only the `<label>` is shown to users.

<Example code={`<div class="text-input-container mb-md">
    <label for="exampleTextInputWithEmptyPlaceholder">Email address</label>
    <input type="email" class="text-input-field" id="exampleTextInputWithEmptyPlaceholder" placeholder="">
  </div>`} />

> If the placeholder text is filled, `<label>`s will automatically adjust to their floated position.

Reviewer: ```suggestion
If a placeholder text is supplied, `<label>`s will automatically adjust to their floated position.
```

      <div class="input-container" data-bs-suffix="€">
        <input type="text" id="inputTextWithSuffix" aria-describedby="suffixTextHelpBlock" class="text-input-field" placeholder="">
      </div>
    </div>
    <div id="suffixTextHelpBlock" class="text-input-helper">
      Enter your price in Euro (€).
    </div>
  </div>
  <div class="text-input">
    <div class="text-input-container mt-md">
      <label for="inputTextWithPrefixSuffix">Price</label>
      <div class="input-container" data-bs-prefix="£" data-bs-suffix=".00">
        <input type="text" id="inputTextWithPrefixSuffix" aria-describedby="prefixSuffixTextHelpBlock" class="text-input-field" placeholder="Enter amount">
      </div>
    </div>
    <div id="prefixSuffixTextHelpBlock" class="text-input-helper">
      Enter your price in Pounds (£) without decimals.
    </div>
  </div>
  <div class="text-input">
    <div class="text-input-container mt-md">
    <svg aria-hidden="true">
      <use xlink:href="/docs/${getConfig().docs_version}/assets/img/ouds-web-sprite.svg#heart-recommend"/>
    </svg>
    <label for="inputTextWithPrefixSuffixLeadingTrailing">Email</label>

>     <div class="input-container" data-bs-prefix="🖳" data-bs-suffix="@orange.com">

Author: It's a sample to display an icon as prefix ;-)

    <input type="email" class="text-input-field" id="exampleTextInputWithPlaceholder" placeholder="name@example.com">
  </div>`} />

When there’s a value already defined, `<label>`s will automatically adjust to their floated position.

<Example code={`<div class="text-input-container mb-md">
    <label for="exampleTextInputWithValue">Input with value</label>
    <input type="email" class="text-input-field" id="exampleTextInputWithValue" placeholder="name@example.com" value="test@example.com">
  </div>`} />

## Variants

### Outlined

Add `.text-input-container-outlined` for a minimalist input with a transparent background and a visible stroke outlining the field. This style may be interesting for contexts other than form pages:
* When inputs need to feel lightweight and unobtrusive
* In a header (search field)
* In a selection/filtering feature in a product catalog

<Example code={`<div class="text-input-container text-input-container-outlined mb-md">
    <label for="exampleTextInputOutlined">Firstname</label>
    <input type="text" class="text-input-field" id="exampleTextInputOutlined" placeholder="">
  </div>`} />

> ### Rounded

Reviewer: This paragraph is probably brand specific (Orange for now, business tools in the future), Sosh and wireframe are rounded by default

Use `<BrandSpecific />`

Author: Ok I enclosed that in a `<BrandSpecific brand="orange">` section

</Callout>

## Overview

Wrap a pair of `<input class="text-input-field">` and `<label>` elements in `.text-input-container` to style OUDS textual form fields. Note that the `<label>` must come before the `<input>`.

A `placeholder` attribute is required on each `<input>` as our CSS-only floating label implementation relies on the `:placeholder-shown` pseudo-element to detect when the input is empty.

If the placeholder text is empty, it is not visible: only the `<label>` is shown to users.

<Callout type="warning">
To be compliant with the WCAG [1.3.5: Identify Input Purpose](https://www.w3.org/WAI/WCAG21/Understanding/identify-input-purpose) criteria, for each input field collecting information about the user, add the appropriate token to the autocomplete attribute (sufficient [technique H98](https://www.w3.org/WAI/WCAG21/Techniques/html/H98)).
</Callout>

<Example code={`<div class="text-input-container mb-md">
    <label for="exampleTextInputWithEmptyPlaceholder">Email address</label>
    <input type="email" autocomplete="email" class="text-input-field" id="exampleTextInputWithEmptyPlaceholder" placeholder="">
  </div>`} />

> If the placeholder text is supplied, `<label>`s will automatically adjust to their floated position.

Reviewer 2: since placeholder is singular, I would probably use singular for labels too, since there is only one per input

Author: Ok fixed

  </div>
  <div class="text-input-container text-input-container-outlined mb-md">
    <svg aria-hidden="true">
      <use xlink:href="${getVersionedDocsPath('/assets/img/ouds-web-sprite.svg#heart-empty')}"/>
    </svg>
    <label for="exampleTextInputAltValueWithIcon">Label</label>
    <input type="email" class="text-input-field" id="exampleTextInputAltValueWithIcon" placeholder="placeholder" value="Input">
  </div>`} />

### Trailing action

Trailing actions can be added to the right side of the input field by placing a minimal icon only button (`.btn`, `.btn-minimal` and `.btn-icon`) inside the `.text-input-container`. It can be used to provide actions related to the field: clear input, toggle password visibility, open a date picker, etc.

<Example code={`<div class="text-input-container mb-md">
    <label for="exampleTextInputEmptyWithAction">Label</label>
    <input type="email" class="text-input-field" id="exampleTextInputEmptyWithAction" placeholder="">
    <button class="btn btn-minimal btn-icon">
      <svg aria-hidden="true">
        <use xlink:href="${getVersionedDocsPath('/assets/img/ouds-web-sprite.svg#heart-empty')}"/>
      </svg>

>       <span class="visually-hidden">Default</span>

Reviewer 2: Maybe we should make more clear that this label should be updated with a good label corresponding to the button goal. Either by using a real case or a text clearly saying, you should update that.

</Callout>

<Example code={`<div class="text-input">
    <div class="text-input-container">
      <label for="exampleTextInputInvalidEmpty">Username</label>
      <input type="text" class="text-input-field is-invalid" id="exampleTextInputInvalidEmpty" aria-describedby="usernameTextHelpBlock" aria-labelledby="usernameFeedback" placeholder="">
    </div>
    <div id="usernameTextHelpBlock" class="text-input-helper">
      Please choose a username.
    </div>
    <div id="usernameFeedback" class="text-input-error">
      Username is required.
    </div>
  </div>
  <div class="text-input">
    <div class="text-input-container mt-md">
      <label for="exampleTextInputInvalid">Email</label>
      <input type="text" class="text-input-field is-invalid" id="exampleTextInputInvalid" value="john.doe@example" aria-labelledby="emailFeedback" placeholder="">
       <button class="btn btn-minimal btn-icon">
        <svg aria-hidden="true">
          <use xlink:href="${getVersionedDocsPath('/assets/img/ouds-web-sprite.svg#heart-empty')}"/>
        </svg>

>         <span class="visually-hidden">Default</span>

Author: Ok fixed
